server/db/models/campaigns.js

```diff
@@ -10,5 +11,76 @@ export default db.Model.extend({
+    const calls = this.relations.calls;
+
+    // some records have duplicate props, this renames the props and returns
+    // a version of the object that no longer has the duplicate prop name.
+    function samePropProcessing(object, objectType, propName, index) {
+      // create a copy because linter likes functional
+      const cloneObject = object;
+      if (index) {
+        // if we're given an index, it's for naming purposes, use it
+        cloneObject[`${objectType}_${index}_${propName}`] = cloneObject[propName];
+      } else {
+        cloneObject[`${objectType}_${propName}`] = cloneObject[propName];
+      }
+      // delete the potential duplicate prop name
+      delete cloneObject[propName];
+      return cloneObject;
+    }
+    // this will be an accumulator that will collect csv headers from record keys
+    let headers = [];
+    // make an array of strings that are lines in the csv.
+    const joined = calls.map((call) => {
+      const responses = call.relations.responses;
+      // this variable will hold the complete record for one line in the csv
+      // but we want to set it conditionally
+      let record;
+      // reassign the cloned call with each processing to keep it clean as we go.
+      let callClone = samePropProcessing(call.attributes, 'call', 'id');
+      callClone = samePropProcessing(callClone, 'call', 'updated_at');
+      callClone = samePropProcessing(callClone, 'call', 'created_at');
+
+      if (responses.length > 0) {
+        // make an array of processed response objects from our array of response models.
+        const responsesClone = responses.models.map((response, index) => {
+          let responseClone = samePropProcessing(response.attributes, 'response', 'id', index + 1);
+          responseClone = samePropProcessing(responseClone, 'response', 'updated_at', index + 1);
+          responseClone = samePropProcessing(responseClone, 'response', 'created_at', index + 1);
+          responseClone = samePropProcessing(responseClone, 'response', 'response', index + 1);
+          responseClone = samePropProcessing(responseClone, 'response', 'question_id', index + 1);
+          // since these will be merged with a call object, we'll already have call_id
+          delete responseClone.call_id;
+          return responseClone;
+        });
+        // add the cloned call to the front of the array before we pass the
+        // array as a set of arguments
+        responsesClone.unshift(callClone);
+        // set our record to a merged version of the processed call and its associated responses.
+        record = Object.assign.apply(null, responsesClone);
+      } else {
+        // if there were no responses the processed call is all we return.
+        record = callClone;
+      }
+      const keys = Object.getOwnPropertyNames(record);
+      // merge record keys into our headers array
+      headers = Object.assign(headers, keys);
```

Comment:
Same as above. In this case .concat works well :)

Reply:
However here we are operating specifically on the array and I appreciate you pointing out the potential for non-index enumerable properties to get in here, so I am going to use lodash.union here.